### PR TITLE
Add an overlay surface on top of embedded UIViews.

### PR DESCRIPTION
**Pull request by amirh**
The overlay surfaces are going to be the same IOSSurface implementation
with the platform views controller set to null (so these are surfaces
that don't support embedding platform views to them).

  * Adds a FlutterOverlayView which is a UIView that's showing an
    overlay surface.
  * Creates an overlay surface for each embedded UIView (done in
    FlutterPlatformViewsController).
  * Changes CompositeEmbeddedView to return a new canvas.
  * Makes the PlatformViewLayer replace the PaintContext's canvas with
    the canvas for the overlay view.
  * Changed canvas in PaintContext to be a pointer so it can be changed.

TBD in following PRs:
  * Copy the current canvas state when replacing a canvas in PaintContext.
  * Make FlutterOverlayView work with a GL backend (currently it only
    works with software rendering)

flutter/flutter#19030

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  @throw([NSException exceptionWithName:@"FlutterOverlayView must initWithDelegate"
+                                 reason:nil
+                               userInfo:nil]);
+}
+
+- (instancetype)init {
+  self = [super initWithFrame:CGRectZero];
+
+  if (self) {
+    self.layer.opaque = NO;
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: Shouldn't you also set the opacity to 0.0?

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    self.userInteractionEnabled = NO;
+  }
+
+  return self;
+}
+
+- (void)layoutSubviews {
+  if ([self.layer isKindOfClass:[CAEAGLLayer class]]) {
+    CAEAGLLayer* layer = reinterpret_cast<CAEAGLLayer*>(self.layer);
+    layer.allowsGroupOpacity = YES;
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: Why allow group opacity? If anything, this should be unset.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 3]
```diff
@@ -124,25 +128,39 @@
...
 
-  UIView* view = views_[view_id];
+  UIView* view = views_[view_id].get();
   [view setFrame:rect];
   composition_order_.push_back(view_id);
+
+  composition_frames_.push_back(
+      overlays_[view_id]->surface->AcquireFrame(params.canvasBaseLayerSize));
+  SkCanvas* canvas = composition_frames_.back()->SkiaCanvas();
+  canvas->clear(SkColorSetARGB(0x00, 0x00, 0x00, 0x00));
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Nit: `SK_ColorTRANSPARENT` is clearer.

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm [hunk 4]
```diff
@@ -124,25 +128,39 @@
...
+  return *canvas;
 }
 
-void FlutterPlatformViewsController::Present() {
+bool FlutterPlatformViewsController::Present() {
+  bool retval = true;
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: Nit: `did_submit` or something more descriptive than "return value".

File: flow/embedded_views.h [hunk 5]
```diff
@@ -25,8 +27,8 @@ class ExternalViewEmbedder {
   ExternalViewEmbedder() = default;
 
   // Must be called on the UI thread.
-  virtual void CompositeEmbeddedView(int view_id,
-                                     const EmbeddedViewParams& params) {}
+  virtual SkCanvas& CompositeEmbeddedView(int view_id,
```
Comment (chinmaygarde) [inline flow/embedded_views.h]: Can this return a pointer instead?

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+#include "flutter/fml/platform/darwin/cf_utils.h"
+#include "flutter/fml/synchronization/waitable_event.h"
+#include "flutter/fml/trace_event.h"
+#include "flutter/shell/common/platform_view.h"
+#include "flutter/shell/common/rasterizer.h"
+#include "flutter/shell/platform/darwin/ios/ios_surface_gl.h"
+#include "flutter/shell/platform/darwin/ios/ios_surface_software.h"
+#include "third_party/skia/include/utils/mac/SkCGUtils.h"
+
+@implementation FlutterOverlayView
```
Comment (chinmaygarde) [inline shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: As written, there seems to be no functional difference between a FlutterView and FlutterOverlayView. Can they me merged with the few differences between the two accounted for by patching the original FlutterView?

Comment (chinmaygarde) [reply shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: TODO is fine.

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm (hunk 1 above)
Comment (amirh) [reply shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: Discussed offline - we need a 1.0 opacity here as this is actually showing flutter rendered contents.

File: flow/embedded_views.h (hunk 5 above)
Comment (amirh) [reply flow/embedded_views.h]: done

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm (hunk 6 above)
Comment (amirh) [reply shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: Discussed offline, added a TODO to revisit this and see if we can merge after we land GL support.

File: shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm (hunk 2 above)
Comment (amirh) [reply shell/platform/darwin/ios/framework/Source/FlutterOverlayView.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 3 above)
Comment (amirh) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

File: shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm (hunk 4 above)
Comment (amirh) [reply shell/platform/darwin/ios/framework/Source/FlutterPlatformViews.mm]: done

**Pull request merged**